### ISSUE TITLE
ResNet18 encrypted inference result is inconsistent with unencrypted 

### ISSUE THREAD

**Issue by ljjsalt** [facebookresearch/CrypTen#419]
Is it normal that encrypted ResNet18 gives inaccurate output? To produce, run the code below, everything else is default.

In[0]:
```python
import torchvision.models as models
model = models.resnet18(pretrained=True)
model.eval()
x = torch.rand(1, 3, 224, 224)
model(torch.rand(1,3,224,224))
```
Out[0]:
```
tensor([[-1.8519e+00,  8.8982e-01,  1.3337e+00,  1.3233e+00,  1.1508e+00,
         -2.5488e-01,  1.0160e+00,  2.9217e-02, -1.2994e+00, -5.3965e-01,
          1.1163e+00,  2.4142e+00,  1.5133e+00,  2.7156e+00,  2.5608e+00,
          1.2186e+00,  1.5630e+00,  2.3750e-01,  1.3247e+00,  1.9385e+00,
          3.2091e-01,  1.8171e+00,  1.6086e+00,  1.5994e+00,  5.6366e-01,
         -5.6269e-03,  2.9198e-01,  3.3615e-01, -3.8973e-01, -1.1169e+00,
         -9.2489e-01,  1.1535e+00, -9.6640e-01,  8.2209e-01,  1.9923e+00,
         -8.5126e-01, -6.2844e-01, -3.0272e-01,  1.1387e+00,  1.4831e-01,
          2.3511e+00,  7.5405e-01,  1.7646e+00,  6.7731e-01,  1.4927e+00,
          2.4418e-01,  1.8786e+00, -3.4501e-01, -2.4688e-01, -3.3947e-01,
          1.4249e+00, -1.1743e+00,  1.3152e+00,  1.0508e+00,  5.0349e-01,
          9.3468e-01,  2.3554e-01,  1.5904e-01,  1.0940e+00,  1.9058e+00,
          1.1053e+00,  6.7723e-01,  6.3189e-03, -8.6582e-01,  5.5522e-01,
          2.1198e+00, -3.3975e-02, -8.9284e-01,  5.2588e-02,  2.3821e+00,
          1.6051e+00,  2.4979e+00,  5.0074e-01,  2.3194e+00,  6.2551e-01,
          2.2954e+00,  8.2466e-01,  1.6867e+00,  3.6946e+00,  3.2785e+00,
         -3.7752e-01,  1.2754e+00,  5.2363e-01, -1.8287e+00, -7.3280e-01,
          1.7236e+00, -2.2116e-01,  2.2446e+00,  8.2453e-01,  2.3115e+00,
         -1.6595e-01,  5.3017e-01,  3.2368e+00,  7.6045e-01,  3.0556e+00,
         -1.4584e-01,  1.4968e+00, -1.4203e+00, -5.9927e-01, -2.1436e-01,
          5.8032e-02, -1.9456e+00, -1.3513e+00,  9.3868e-01, -2.2638e-01,
         -1.1066e+00, -1.4399e+00,  2.3487e+00, -1.7167e+00, -2.9520e-01,
         -1.0051e+00,  5.6177e+00,  1.8110e+00,  1.1170e+00,  1.7341e+00,
         -1.7251e+00, -1.4984e+00,  7.1827e-01, -3.3371e-01, -8.9697e-01,
          2.4247e-01, -1.6477e+00, -5.9715e-01, -9.6308e-01,  5.8787e-01,
...
          1.8035e+00,  5.6114e-01, -1.5724e+00,  2.6077e-01, -2.6800e-01,
          6.9415e-02, -5.6132e-01,  4.5036e-01,  9.8340e-01,  3.5447e-01,
         -1.5842e-01, -1.1212e+00, -1.5008e+00, -2.6297e+00, -2.0073e+00,
         -1.5612e+00, -2.0720e+00, -7.2153e-01,  2.0094e-01,  2.5504e+00]],
       grad_fn=<AddmmBackward>)
```
In[1]:
```python
import crypten.mpc as mpc
import crypten.communicator as comm


@mpc.run_multiprocess(world_size=2)
def encrypt_model_and_data():
    # Encrypt model from Alice 
    dummy_input = torch.empty((1, 3, 224, 224))
    private_model = crypten.nn.from_pytorch(model, dummy_input)
    private_model.encrypt(src=ALICE)
    
    # Load data to Bob
    data_enc = crypten.cryptensor(x, src=BOB)

    # Classify the encrypted data
    private_model.eval()
    output_enc = private_model(data_enc)
    
    output = output_enc.get_plain_text()
    print(output)
    
encrypt_model_and_data()
```
Out[1]:
```
tensor([[-1.4371e+00,  6.3200e-01,  1.3528e+00,  1.3563e+00,  1.2632e+00,
         -3.6163e-02,  1.3134e+00,  3.4497e-01, -9.8889e-01, -6.4621e-01,
          5.0647e-01,  1.5046e+00,  7.2696e-01,  2.0657e+00,  1.8627e+00,
          7.6585e-01,  9.3190e-01, -2.1243e-01,  1.1035e+00,  1.1885e+00,
          4.4688e-01,  1.3081e+00,  1.3365e+00,  1.1313e+00, -1.5125e-01,
          1.0938e-01,  2.5414e-01,  5.0507e-01, -1.7923e-01, -1.1130e+00,
         -5.9970e-01,  6.4502e-01, -5.5766e-01,  1.1828e+00,  2.6588e+00,
         -4.2456e-01, -4.3655e-01,  1.0643e-01,  1.2305e+00,  2.4214e-01,
          1.9193e+00,  1.0918e+00,  1.6455e+00,  4.5152e-01,  1.5556e+00,
          6.3084e-01,  1.8218e+00, -6.3753e-01,  2.7673e-01,  1.2111e-01,
          1.6111e+00, -1.2622e+00,  1.5854e+00,  1.2860e+00,  7.3611e-01,
          9.1612e-01,  5.7811e-01,  2.5621e-01,  1.5110e+00,  1.5306e+00,
          1.2794e+00,  7.9021e-01,  1.1850e-01, -3.4750e-01,  3.4093e-01,
          2.4651e+00,  2.7513e-01, -3.3707e-01,  5.3188e-01,  2.3792e+00,
          1.4732e+00,  2.6320e+00,  7.5348e-02,  2.1656e+00,  3.0827e-01,
          2.3345e+00,  9.0842e-01,  1.8802e+00,  3.5430e+00,  3.4385e+00,
         -5.8148e-01,  1.5122e+00,  5.5257e-01, -1.6625e+00, -3.5339e-01,
          1.4687e+00, -2.5038e-01,  1.9832e+00,  2.8864e-01,  1.8558e+00,
         -5.3981e-01,  7.3822e-02,  2.6862e+00,  4.2509e-01,  2.3840e+00,
         -9.0164e-01,  1.0351e+00, -7.6277e-01,  1.8768e-03,  3.2994e-01,
          3.8773e-01, -1.9980e+00, -7.0514e-01,  1.0517e+00,  1.8356e-01,
         -1.3201e+00, -1.0684e+00,  1.9585e+00, -1.4611e+00, -3.7321e-01,
         -1.2294e+00,  4.9744e+00,  1.9875e+00,  9.8425e-01,  1.7598e+00,
         -1.7744e+00, -1.4577e+00,  4.7047e-01, -4.2374e-02, -3.2822e-01,
          1.0087e+00, -1.5399e+00, -4.7879e-01, -5.8801e-01,  1.1493e+00,
...
          1.6823e+00,  3.4715e-01, -1.7715e+00,  1.6621e-01, -6.1606e-01,
         -4.6448e-02, -9.8816e-01,  3.6690e-01,  8.4319e-01,  1.9211e-02,
          2.2021e-01, -1.4497e+00, -1.7088e+00, -2.6095e+00, -1.9857e+00,
         -1.6237e+00, -2.2445e+00, -8.6095e-01,  3.5553e-02,  2.3689e+00]])
```

**Comment by lvdmaaten** [comment]
These numbers seem fairly close, right? When you do inference in PyTorch, you are performing the computations using 32-but floating-point numbers. By contrast, CrypTen uses 16-bit fixed-point numbers to do the computations. So small numerical deviations are to be expected, in particular, in models that perform many operations like ResNets.

**Comment by ljjsalt** [comment]
Hi @lvdmaaten , does this imply if the model was trained under `fp16`, then there should be no error in the encrypted output? Since I am worried about the accuracy lose.

**Comment by lvdmaaten** [comment]
There would still be an error: note that `fp16` is an 16-bit *floating-point* representations, whereas CrypTen uses a 16-bit *fixed-point* representation. The two are not equivalent (besides, there are multiple 16-bit floating-point number representations in use: for example, `bfloat` is different from `fb16`). Also note that you can [change the number of bits used](https://github.com/facebookresearch/CrypTen/blob/main/configs/default.yaml#L7) but there are trade-offs here between the fidelity of the number representation and the probability of wrap-around errors because the ring side is fixed to `2**64`.

**Comment by ljjsalt** [comment]
Hi @lvdmaaten, thanks for explanation again. I just tested my own model which uses polynomial for activations, and I observed that my accuracy drops from 75% to 71%. Is it possible to improve the accuracy?

**Comment by lvdmaaten** [comment]
I haven't tried this very extensively, but I think fully encrypted finetuning with a small learning rate may help.

**Comment by ljjsalt** [comment]
> I haven't tried this very extensively, but I think fully encrypted finetuning with a small learning rate may help.

Thanks, I will consider this.
